misc/futures-bounded/src/unique.rs

```diff
@@ -0,0 +1,218 @@
+use std::collections::HashSet;
+use std::future::Future;
+use std::hash::Hash;
+use std::task::{Context, Poll, Waker};
+use std::time::Duration;
+
+use futures_timer::Delay;
+use futures_util::future::{select, BoxFuture, Either};
+use futures_util::stream::FuturesUnordered;
+use futures_util::{ready, FutureExt, StreamExt};
+
+use crate::{PushResult, Timeout};
+
+/// Represents a set of (Worker)-[Future]s.
+///
+/// This wraps [FuturesUnordered] but bounds it by time and size. Every worker has to have a unique ID.
+/// In other words, each worker must finish within the specified time and the set never outgrows its capacity.
+pub struct UniqueWorkers<ID, O> {
+    timeout: Duration,
+    capacity: usize,
+    keys: HashSet<ID>,
+    inner: FuturesUnordered<BoxFuture<'static, (ID, Result<O, Timeout>)>>,
+
+    empty_waker: Option<Waker>,
+    full_waker: Option<Waker>,
+}
+
+impl<ID, O> UniqueWorkers<ID, O> {
+    pub fn new(timeout: Duration, capacity: usize) -> Self {
+        Self {
+            timeout,
+            capacity,
+            keys: HashSet::with_capacity(capacity),
+            inner: Default::default(),
+            empty_waker: None,
+            full_waker: None,
+        }
+    }
+}
+
+impl<ID, O> UniqueWorkers<ID, O>
+where
+    ID: Clone + Hash + Eq + Send + 'static,
+{
+    /// Push a worker into the set.
+    /// This method adds the given worker with defined `worker_id` to the set and returns [PushResult::Ok].
+    /// The result [PushResult::ExistedID] says that the set already contains passed `worker_id`.
+    /// If length of the set is equal the capacity, this method returns [PushResult::BeyondCapacity].
+    /// The result different from [PushResult::Ok] means that worker was not added to the set.
+    pub fn try_push<F>(&mut self, worker_id: ID, worker: F) -> PushResult
+    where
+        F: Future<Output = O> + Send + 'static + Unpin,
+    {
+        // Check worker's ID
+        if !self.keys.insert(worker_id.clone()) {
+            return PushResult::ExistedID;
```

Comment:
Should we return the worker instead? I think it would be a better API if we leave it up to the user to drop it if they don't need it.

Reply:
ok